File: backend/src/main/java/com/ultrabms/entity/PDC.java
package com.ultrabms.entity;

import com.ultrabms.entity.enums.NewPaymentMethod;
import com.ultrabms.entity.enums.PDCStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * PDC (Post-Dated Cheque) entity representing cheques received from tenants.
 * Tracks PDC lifecycle from receipt through deposit to clearance or bounce.
 *
 * Story 6.3: Post-Dated Cheque (PDC) Management
 */
@Entity
@Table(
    name = "pdcs",
    uniqueConstraints = {
        @UniqueConstraint(
            name = "uk_pdc_cheque_tenant",
            columnNames = {"cheque_number", "tenant_id"}
        )
    },
    indexes = {
        @Index(name = "idx_pdcs_tenant_id", columnList = "tenant_id"),
        @Index(name = "idx_pdcs_status", columnList = "status"),
        @Index(name = "idx_pdcs_cheque_date", columnList = "cheque_date"),
        @Index(name = "idx_pdcs_cheque_number", columnList = "cheque_number"),
        @Index(name = "idx_pdcs_deposit_date", columnList = "deposit_date"),
        @Index(name = "idx_pdcs_invoice_id", columnList = "invoice_id"),
        @Index(name = "idx_pdcs_lease_id", columnList = "lease_id"),
        @Index(name = "idx_pdcs_bank_name", columnList = "bank_name"),
        @Index(name = "idx_pdcs_status_cheque_date", columnList = "status, cheque_date")
    }
)
@Data
@Builder
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
public class PDC extends BaseEntity {

    // =================================================================
    // CHEQUE IDENTIFICATION
    // =================================================================

    /**
     * Cheque number from the physical cheque
     * Must be unique per tenant
     */
    @NotBlank(message = "Cheque number cannot be blank")
    @Size(min = 3, max = 50, message = "Cheque number must be between 3 and 50 characters")
    @Column(name = "cheque_number", nullable = false, length = 50)
    private String chequeNumber;

    /**
     * Bank name that issued the cheque
     */
    @NotBlank(message = "Bank name cannot be blank")
    @Size(max = 100, message = "Bank name must be less than 100 characters")
    @Column(name = "bank_name", nullable = false, length = 100)
    private String bankName;

    // =================================================================
    // TENANT AND RELATIONSHIPS
    // =================================================================

    /**
     * Tenant who issued this PDC
     */
    @NotNull(message = "Tenant cannot be null")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "tenant_id", nullable = false)
    private Tenant tenant;

    /**
     * Invoice this PDC is linked to (optional)
     * Used for automatic payment recording on clear
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "invoice_id")
    private Invoice invoice;

    /**
     * Lease ID associated with this PDC (optional)
     */
    @Column(name = "lease_id")
    private UUID leaseId;

    // =================================================================
    // AMOUNT AND DATES
    // =================================================================

    /**
     * Cheque amount (AED)
     */
    @NotNull(message = "Amount cannot be null")
    @DecimalMin(value = "0.01", message = "Amount must be greater than 0")
    @Column(name = "amount", precision = 12, scale = 2, nullable = false)
    private BigDecimal amount;

    /**
     * Date printed on the cheque (post-dated)
     */
    @NotNull(message = "Cheque date is required")
    @Column(name = "cheque_date", nullable = false)
    private LocalDate chequeDate;

    /**
     * Date when PDC was deposited to bank
     */
    @Column(name = "deposit_date")
    private LocalDate depositDate;

    /**
     * Date when payment was confirmed/cleared
     */
    @Column(name = "cleared_date")
    private LocalDate clearedDate;

    /**
     * Date when cheque bounced
     */
    @Column(name = "bounced_date")
    private LocalDate bouncedDate;

    /**
     * Date when PDC was withdrawn/returned to tenant
     */
    @Column(name = "withdrawal_date")
    private LocalDate withdrawalDate;

    // =================================================================
    // STATUS AND TRACKING
    // =================================================================

    /**
     * Current PDC status
     */
    @NotNull(message = "Status cannot be null")
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    @Builder.Default
    private PDCStatus status = PDCStatus.RECEIVED;

    /**
     * Reason for bounce (when status = BOUNCED)
     */
    @Size(max = 255, message = "Bounce reason must be less than 255 characters")
    @Column(name = "bounce_reason", length = 255)
    private String bounceReason;

    /**
     * Reason for withdrawal (when status = WITHDRAWN)
     */
    @Size(max = 255, message = "Withdrawal reason must be less than 255 characters")
    @Column(name = "withdrawal_reason", length = 255)
    private String withdrawalReason;

    /**
     * New payment method when PDC is withdrawn
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "new_payment_method", length = 20)
    private NewPaymentMethod newPaymentMethod;

    /**
     * Transaction ID for alternative payment (when withdrawn with bank transfer)
     */
    @Size(max = 100, message = "Transaction ID must be less than 100 characters")
    @Column(name = "transaction_id", length = 100)
    private String transactionId;

    // =================================================================
    // REPLACEMENT CHAIN
    // =================================================================

    /**
     * Reference to the replacement PDC (when this PDC is REPLACED)
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "replacement_pdc_id")
    private PDC replacementPdc;

    /**
     * Reference to the original bounced PDC (when this is a replacement)
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "original_pdc_id")
    private PDC originalPdc;

    // =================================================================
    // BANK ACCOUNT FOR DEPOSIT
    // =================================================================

    /**
     * Bank account ID where PDC was deposited (from Story 6.5)
     * Optional until Story 6.5 is implemented
     */
    @Column(name = "bank_account_id")
    private UUID bankAccountId;

    // =================================================================
    // NOTES AND METADATA
    // =================================================================

    /**
     * General notes about the PDC
     */
    @Size(max = 500, message = "Notes must be less than 500 characters")
    @Column(name = "notes", length = 500)
    private String notes;

    /**
     * User who registered this PDC
     */
    @NotNull(message = "Created by cannot be null")
    @Column(name = "created_by", nullable = false)
    private UUID createdBy;

    // =================================================================
    // LIFECYCLE CALLBACKS
    // =================================================================

    /**
     * Pre-persist callback to set default values
     */
    @PrePersist
    protected void onCreate() {
        if (this.status == null) {
            this.status = PDCStatus.RECEIVED;
        }
    }

    // =================================================================
    // STATUS TRANSITION METHODS
    // =================================================================

    /**
     * Check if PDC can be deposited
     * Only DUE status PDCs can be deposited
     */
    public boolean canBeDeposited() {
        return this.status == PDCStatus.DUE;
    }

    /**
     * Check if PDC can be cleared
     * Only DEPOSITED status PDCs can be cleared
     */
    public boolean canBeCleared() {
        return this.status == PDCStatus.DEPOSITED;
    }

    /**
     * Check if PDC can be bounced
     * Only DEPOSITED status PDCs can be marked as bounced
     */
    public boolean canBeBounced() {
        return this.status == PDCStatus.DEPOSITED;
    }

    /**
     * Check if PDC can be replaced
     * Only BOUNCED status PDCs can be replaced
     */
    public boolean canBeReplaced() {
        return this.status == PDCStatus.BOUNCED;
    }

    /**
     * Check if PDC can be withdrawn
     * Only RECEIVED or DUE status PDCs can be withdrawn
     */
    public boolean canBeWithdrawn() {
        return this.status == PDCStatus.RECEIVED || this.status == PDCStatus.DUE;
    }

    /**
     * Check if PDC can be cancelled
     * Only RECEIVED status PDCs can be cancelled
     */
    public boolean canBeCancelled() {
        return this.status == PDCStatus.RECEIVED;
    }

    /**
     * Check if PDC is in a final state
     */
    public boolean isInFinalState() {
        return this.status == PDCStatus.CLEARED
            || this.status == PDCStatus.CANCELLED
            || this.status == PDCStatus.REPLACED
            || this.status == PDCStatus.WITHDRAWN;
    }

    /**
     * Check if cheque date is within due window (7 days from today)
     */
    public boolean isWithinDueWindow() {
        if (this.chequeDate == null) {
            return false;
        }
        LocalDate today = LocalDate.now();
        LocalDate dueWindow = today.plusDays(7);
        return !this.chequeDate.isBefore(today) && !this.chequeDate.isAfter(dueWindow);
    }

    /**
     * Check if cheque date is past
     */
    public boolean isChequeOverdue() {
        if (this.chequeDate == null) {
            return false;
        }
        return this.chequeDate.isBefore(LocalDate.now());
    }

    /**
     * Transition to DUE status (automated by scheduler)
     */
    public void transitionToDue() {
        if (this.status != PDCStatus.RECEIVED) {
            throw new IllegalStateException("Cannot transition to DUE from status: " + this.status);
        }
        if (!isWithinDueWindow()) {
            throw new IllegalStateException("Cheque date is not within due window");
        }
        this.status = PDCStatus.DUE;
    }

    /**
     * Mark PDC as deposited
     */
    public void deposit(LocalDate depositDate, UUID bankAccountId) {
        if (!canBeDeposited()) {
            throw new IllegalStateException("PDC cannot be deposited in current status: " + this.status);
        }
        if (depositDate == null) {
            throw new IllegalArgumentException("Deposit date cannot be null");
        }
        this.depositDate = depositDate;
        this.bankAccountId = bankAccountId;
        this.status = PDCStatus.DEPOSITED;
    }

    /**
     * Mark PDC as cleared
     */
    public void clear(LocalDate clearedDate) {
        if (!canBeCleared()) {
            throw new IllegalStateException("PDC cannot be cleared in current status: " + this.status);
        }
        if (clearedDate == null) {
            throw new IllegalArgumentException("Cleared date cannot be null");
        }
        this.clearedDate = clearedDate;
        this.status = PDCStatus.CLEARED;
    }

    /**
     * Mark PDC as bounced
     */
    public void bounce(LocalDate bouncedDate, String bounceReason) {
        if (!canBeBounced()) {
            throw new IllegalStateException("PDC cannot be bounced in current status: " + this.status);
        }
        if (bouncedDate == null) {
            throw new IllegalArgumentException("Bounced date cannot be null");
        }
        if (bounceReason == null || bounceReason.isBlank()) {
            throw new IllegalArgumentException("Bounce reason cannot be blank");
        }
        this.bouncedDate = bouncedDate;
        this.bounceReason = bounceReason;
        this.status = PDCStatus.BOUNCED;
    }

    /**
     * Mark PDC as replaced (links to replacement PDC)
     */
    public void markAsReplaced(PDC replacementPdc) {
        if (!canBeReplaced()) {
            throw new IllegalStateException("PDC cannot be replaced in current status: " + this.status);
        }
        if (replacementPdc == null) {
            throw new IllegalArgumentException("Replacement PDC cannot be null");
        }
        this.replacementPdc = replacementPdc;
        this.status = PDCStatus.REPLACED;
    }

    /**
     * Withdraw PDC (return to tenant)
     */
    public void withdraw(LocalDate withdrawalDate, String withdrawalReason,
                         NewPaymentMethod newPaymentMethod, String transactionId) {
        if (!canBeWithdrawn()) {
            throw new IllegalStateException("PDC cannot be withdrawn in current status: " + this.status);
        }
        if (withdrawalDate == null) {
            throw new IllegalArgumentException("Withdrawal date cannot be null");
        }
        if (withdrawalReason == null || withdrawalReason.isBlank()) {
            throw new IllegalArgumentException("Withdrawal reason cannot be blank");
        }
        this.withdrawalDate = withdrawalDate;
        this.withdrawalReason = withdrawalReason;
        this.newPaymentMethod = newPaymentMethod;
        this.transactionId = transactionId;
        this.status = PDCStatus.WITHDRAWN;
    }

    /**
     * Cancel PDC
     */
    public void cancel() {
        if (!canBeCancelled()) {
            throw new IllegalStateException("PDC cannot be cancelled in current status: " + this.status);
        }
        this.status = PDCStatus.CANCELLED;
    }
}

File: frontend/src/types/pdc.ts
/**
 * PDC (Post-Dated Cheque) Management Types and Interfaces
 * Story 6.3: Post-Dated Cheque (PDC) Management
 */

// ============================================================================
// ENUMS
// ============================================================================

/**
 * PDC status enum
 * Tracks PDC lifecycle from receipt to final state
 * State machine: RECEIVED → DUE → DEPOSITED → CLEARED/BOUNCED/WITHDRAWN
 *                                  BOUNCED → REPLACED
 */
export enum PDCStatus {
  RECEIVED = 'RECEIVED',
  DUE = 'DUE',
  DEPOSITED = 'DEPOSITED',
  CLEARED = 'CLEARED',
  BOUNCED = 'BOUNCED',
  CANCELLED = 'CANCELLED',
  REPLACED = 'REPLACED',
  WITHDRAWN = 'WITHDRAWN'
}

/**
 * New payment method enum for PDC withdrawal replacement
 * When a PDC is withdrawn, tenant may provide alternative payment
 */
export enum NewPaymentMethod {
  BANK_TRANSFER = 'BANK_TRANSFER',
  CASH = 'CASH',
  NEW_CHEQUE = 'NEW_CHEQUE'
}

// ============================================================================
// MAIN INTERFACES
// ============================================================================

/**
 * Full PDC entity
 * Complete post-dated cheque information from backend
 */
export interface PDC {
  id: string;
  chequeNumber: string;
  tenantId: string;
  tenantName: string;
  tenantEmail?: string;
  invoiceId?: string;
  invoiceNumber?: string;
  leaseId?: string;
  leaseNumber?: string;
  bankName: string;
  amount: number;
  chequeDate: string;
  depositDate?: string;
  clearedDate?: string;
  bouncedDate?: string;
  bounceReason?: string;
  status: PDCStatus;
  withdrawalDate?: string;
  withdrawalReason?: string;
  newPaymentMethod?: NewPaymentMethod;
  replacementChequeId?: string;
  replacementChequeNumber?: string;
  originalChequeId?: string;
  originalChequeNumber?: string;
  bankAccountId?: string;
  bankAccountName?: string;
  notes?: string;
  createdBy: string;
  createdByName?: string;
  createdAt: string;
  updatedAt: string;
}

/**
 * PDC list item for table view
 * Minimal fields for efficient list rendering
 */
export interface PDCListItem {
  id: string;
  chequeNumber: string;
  tenantId: string;
  tenantName: string;
  bankName: string;
  amount: number;
  chequeDate: string;
  status: PDCStatus;
  depositDate?: string;
  clearedDate?: string;
  bouncedDate?: string;
  isDue: boolean;
  daysUntilDue?: number;
}

/**
 * PDC detail with full relations
 * Extended PDC info for detail page
 */
export interface PDCDetail extends PDC {
  tenant: {
    id: string;
    fullName: string;
    email: string;
    phone?: string;
  };
  invoice?: {
    id: string;
    invoiceNumber: string;
    totalAmount: number;
    balanceAmount: number;
  };
  lease?: {
    id: string;
    unitNumber: string;
    propertyName: string;
    startDate: string;
    endDate: string;
  };
  bankAccount?: {
    id: string;
    bankName: string;
    maskedAccountNumber: string;
  };
  statusHistory: PDCStatusTransition[];
  replacementChain?: PDCReplacementChainItem[];
}

/**
 * PDC status transition for timeline
 */
export interface PDCStatusTransition {
  fromStatus: PDCStatus | null;
  toStatus: PDCStatus;
  transitionDate: string;
  performedBy: string;
  performedByName: string;
  notes?: string;
}

/**
 * PDC replacement chain item for tracking bounced cheque replacements
 */
export interface PDCReplacementChainItem {
  id: string;
  chequeNumber: string;
  amount: number;
  chequeDate: string;
  status: PDCStatus;
  bouncedDate?: string;
  bounceReason?: string;
}

/**
 * PDC holder information from company profile
 */
export interface PDCHolder {
  companyName: string;
  legalCompanyName?: string;
  tradeLicenseNumber?: string;
}

/**
 * Bank account for PDC deposit selection
 */
export interface BankAccountOption {
  id: string;
  bankName: string;
  maskedAccountNumber: string;
  displayName: string;
}

// ============================================================================
// DTO INTERFACES
// ============================================================================

/**
 * Single cheque entry for PDC registration
 */
export interface PDCChequeEntry {
  chequeNumber: string;
  bankName: string;
  amount: number;
  chequeDate: string;
}

/**
 * Request DTO for creating single PDC
 */
export interface PDCCreateRequest {
  tenantId: string;
  leaseId?: string;
  invoiceId?: string;
  chequeNumber: string;
  bankName: string;
  amount: number;
  chequeDate: string;
  notes?: string;
}

/**
 * Request DTO for bulk PDC registration
 */
export interface PDCBulkCreateRequest {
  tenantId: string;
  leaseId?: string;
  cheques: PDCChequeEntry[];
}

/**
 * Request DTO for depositing PDC
 */
export interface PDCDepositRequest {
  depositDate: string;
  bankAccountId: string;
}

/**
 * Request DTO for clearing PDC
 */
export interface PDCClearRequest {
  clearedDate: string;
}

/**
 * Request DTO for reporting PDC bounce
 */
export interface PDCBounceRequest {
  bouncedDate: string;
  bounceReason: string;
}

/**
 * Request DTO for replacing bounced PDC
 */
export interface PDCReplaceRequest {
  newChequeNumber: string;
  bankName: string;
  amount: number;
  chequeDate: string;
  notes?: string;
}

/**
 * Transaction details for bank transfer withdrawal replacement
 */
export interface WithdrawalTransactionDetails {
  amount: number;
  transactionId: string;
  bankAccountId: string;
}

/**
 * Request DTO for withdrawing PDC
 */
export interface PDCWithdrawRequest {
  withdrawalDate: string;
  withdrawalReason: string;
  newPaymentMethod?: NewPaymentMethod;
  transactionDetails?: WithdrawalTransactionDetails;
}

/**
 * Filter parameters for PDC list
 */
export interface PDCFilter {
  search?: string;
  status?: PDCStatus | PDCStatus[] | 'ALL';
  tenantId?: string;
  bankName?: string;
  fromDate?: string;
  toDate?: string;
  leaseId?: string;
  invoiceId?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

/**
 * Filter parameters for PDC withdrawal history
 */
export interface PDCWithdrawalFilter {
  search?: string;
  withdrawalReason?: string | 'ALL';
  fromDate?: string;
  toDate?: string;
  page?: number;
  size?: number;
  sortBy?: string;
  sortDirection?: 'ASC' | 'DESC';
}

// ============================================================================
// DASHBOARD TYPES
// ============================================================================

/**
 * PDC Dashboard KPI data
 */
export interface PDCDashboardKPIs {
  pdcsDueThisWeek: {
    count: number;
    totalValue: number;
  };
  pdcsDeposited: {
    count: number;
    totalValue: number;
  };
  totalOutstandingValue: number;
  recentlyBouncedCount: number;
}

/**
 * Full PDC dashboard data
 */
export interface PDCDashboard {
  pdcsDueThisWeek: {
    count: number;
    totalValue: number;
  };
  pdcsDeposited: {
    count: number;
    totalValue: number;
  };
  totalOutstandingValue: number;
  recentlyBouncedCount: number;
  upcomingPDCs: PDCListItem[];
  recentlyDepositedPDCs: PDCListItem[];
}

/**
 * Tenant PDC history with bounce rate
 */
export interface TenantPDCHistory {
  tenantId: string;
  tenantName: string;
  totalPDCs: number;
  clearedPDCs: number;
  bouncedPDCs: number;
  pendingPDCs: number;
  bounceRate: number;
  pdcs: PDCListItem[];
}

// ============================================================================
// WITHDRAWAL HISTORY TYPES
// ============================================================================

/**
 * PDC withdrawal history item
 */
export interface PDCWithdrawalHistoryItem {
  id: string;
  originalChequeNumber: string;
  tenantId: string;
  tenantName: string;
  withdrawalDate: string;
  amount: number;
  withdrawalReason: string;
  newPaymentMethod?: NewPaymentMethod;
  associatedPDCId?: string;
  associatedChequeNumber?: string;
  transactionId?: string;
}

// ============================================================================
// API RESPONSE TYPES
// ============================================================================

/**
 * Response from create PDC endpoint
 */
export interface CreatePDCResponse {
  success: boolean;
  message: string;
  data: PDC;
  timestamp: string;
}

/**
 * Response from bulk create PDCs endpoint
 */
export interface BulkCreatePDCResponse {
  success: boolean;
  message: string;
  data: PDC[];
  timestamp: string;
}

/**
 * Response from get PDC endpoint
 */
export interface GetPDCResponse {
  success: boolean;
  message: string;
  data: PDCDetail;
  timestamp: string;
}

/**
 * Response from list PDCs endpoint
 */
export interface PDCListResponse {
  success: boolean;
  message: string;
  data: {
    content: PDCListItem[];
    totalElements: number;
    totalPages: number;
    page: number;
    size: number;
  };
  timestamp: string;
}

/**
 * Response from PDC dashboard endpoint
 */
export interface PDCDashboardResponse {
  success: boolean;
  message: string;
  data: PDCDashboard;
  timestamp: string;
}

/**
 * Response from PDC status action endpoints (deposit, clear, bounce, withdraw, cancel)
 */
export interface PDCStatusActionResponse {
  success: boolean;
  message: string;
  data: PDC;
  timestamp: string;
}

/**
 * Response from replace PDC endpoint
 */
export interface PDCReplaceResponse {
  success: boolean;
  message: string;
  data: {
    original: PDC;
    replacement: PDC;
  };
  timestamp: string;
}

/**
 * Response from tenant PDC history endpoint
 */
export interface TenantPDCHistoryResponse {
  success: boolean;
  message: string;
  data: {
    content: PDCListItem[];
    totalElements: number;
    totalPages: number;
    page: number;
    size: number;
    bounceRate: number;
  };
  timestamp: string;
}

/**
 * Response from PDC withdrawal history endpoint
 */
export interface PDCWithdrawalHistoryResponse {
  success: boolean;
  message: string;
  data: {
    content: PDCWithdrawalHistoryItem[];
    totalElements: number;
    totalPages: number;
    page: number;
    size: number;
  };
  timestamp: string;
}

/**
 * Response from bank accounts endpoint for PDC deposit
 */
export interface BankAccountsResponse {
  success: boolean;
  message: string;
  data: BankAccountOption[];
  timestamp: string;
}

/**
 * Response from PDC holder endpoint (company profile)
 */
export interface PDCHolderResponse {
  success: boolean;
  message: string;
  data: PDCHolder;
  timestamp: string;
}

// ============================================================================
// HELPER TYPES
// ============================================================================

/**
 * PDC status display information
 */
export interface PDCStatusInfo {
  value: PDCStatus;
  label: string;
  color: 'gray' | 'amber' | 'blue' | 'green' | 'red' | 'purple' | 'orange' | 'slate';
  icon: string;
  description: string;
}

/**
 * PDC status options for dropdowns and badges
 */
export const PDC_STATUS_OPTIONS: PDCStatusInfo[] = [
  {
    value: PDCStatus.RECEIVED,
    label: 'Received',
    color: 'gray',
    icon: 'inbox',
    description: 'PDC received from tenant'
  },
  {
    value: PDCStatus.DUE,
    label: 'Due',
    color: 'amber',
    icon: 'calendar-clock',
    description: 'Cheque date within 7 days'
  },
  {
    value: PDCStatus.DEPOSITED,
    label: 'Deposited',
    color: 'blue',
    icon: 'building-2',
    description: 'Submitted to bank'
  },
  {
    value: PDCStatus.CLEARED,
    label: 'Cleared',
    color: 'green',
    icon: 'check-circle',
    description: 'Payment confirmed by bank'
  },
  {
    value: PDCStatus.BOUNCED,
    label: 'Bounced',
    color: 'red',
    icon: 'alert-triangle',
    description: 'Payment failed'
  },
  {
    value: PDCStatus.CANCELLED,
    label: 'Cancelled',
    color: 'slate',
    icon: 'x-circle',
    description: 'PDC voided'
  },
  {
    value: PDCStatus.REPLACED,
    label: 'Replaced',
    color: 'purple',
    icon: 'refresh-cw',
    description: 'Replaced with new PDC'
  },
  {
    value: PDCStatus.WITHDRAWN,
    label: 'Withdrawn',
    color: 'orange',
    icon: 'undo-2',
    description: 'Returned to tenant'
  }
];

/**
 * New payment method display information
 */
export interface NewPaymentMethodInfo {
  value: NewPaymentMethod;
  label: string;
  icon: string;
}

/**
 * New payment method options for withdrawal modal
 */
export const NEW_PAYMENT_METHOD_OPTIONS: NewPaymentMethodInfo[] = [
  { value: NewPaymentMethod.BANK_TRANSFER, label: 'Bank Transfer', icon: 'landmark' },
  { value: NewPaymentMethod.CASH, label: 'Cash', icon: 'banknote' },
  { value: NewPaymentMethod.NEW_CHEQUE, label: 'New Cheque', icon: 'file-text' }
];

/**
 * Common bounce reasons for dropdown
 */
export const BOUNCE_REASON_OPTIONS: string[] = [
  'Insufficient Funds',
  'Signature Mismatch',
  'Account Closed',
  'Payment Stopped',
  'Post-dated',
  'Stale Check',
  'Refer to Drawer',
  'Account Frozen',
  'Other'
];

/**
 * Common withdrawal reasons for dropdown
 */
export const WITHDRAWAL_REASON_OPTIONS: string[] = [
  'Cheque Bounced',
  'Replacement Requested',
  'Early Contract Termination',
  'Payment Method Change',
  'Tenant Request',
  'Other'
];

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Get status badge color class
 */
export function getPDCStatusColor(status: PDCStatus): string {
  switch (status) {
    case PDCStatus.RECEIVED:
      return 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-300';
    case PDCStatus.DUE:
      return 'bg-amber-100 text-amber-800 dark:bg-amber-900 dark:text-amber-300';
    case PDCStatus.DEPOSITED:
      return 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300';
    case PDCStatus.CLEARED:
      return 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-300';
    case PDCStatus.BOUNCED:
      return 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-300';
    case PDCStatus.CANCELLED:
      return 'bg-slate-100 text-slate-800 dark:bg-slate-700 dark:text-slate-300';
    case PDCStatus.REPLACED:
      return 'bg-purple-100 text-purple-800 dark:bg-purple-900 dark:text-purple-300';
    case PDCStatus.WITHDRAWN:
      return 'bg-orange-100 text-orange-800 dark:bg-orange-900 dark:text-orange-300';
    default:
      return 'bg-gray-100 text-gray-800';
  }
}

/**
 * Get status badge variant for shadcn Badge
 */
export function getPDCStatusVariant(status: PDCStatus): 'default' | 'secondary' | 'destructive' | 'outline' {
  switch (status) {
    case PDCStatus.CLEARED:
      return 'default';
    case PDCStatus.BOUNCED:
      return 'destructive';
    case PDCStatus.CANCELLED:
    case PDCStatus.REPLACED:
    case PDCStatus.WITHDRAWN:
      return 'secondary';
    default:
      return 'outline';
  }
}

/**
 * Get PDC status label
 */
export function getPDCStatusLabel(status: PDCStatus): string {
  const option = PDC_STATUS_OPTIONS.find(opt => opt.value === status);
  return option?.label ?? status;
}

/**
 * Get new payment method label
 */
export function getNewPaymentMethodLabel(method: NewPaymentMethod): string {
  const option = NEW_PAYMENT_METHOD_OPTIONS.find(opt => opt.value === method);
  return option?.label ?? method;
}

/**
 * Check if PDC can be deposited
 * Only DUE status PDCs can be deposited
 */
export function canDepositPDC(status: PDCStatus): boolean {
  return status === PDCStatus.DUE;
}

/**
 * Check if PDC can be cleared
 * Only DEPOSITED status PDCs can be cleared
 */
export function canClearPDC(status: PDCStatus): boolean {
  return status === PDCStatus.DEPOSITED;
}

/**
 * Check if PDC can be bounced
 * Only DEPOSITED status PDCs can be marked as bounced
 */
export function canBouncePDC(status: PDCStatus): boolean {
  return status === PDCStatus.DEPOSITED;
}

/**
 * Check if PDC can be replaced
 * Only BOUNCED status PDCs can be replaced
 */
export function canReplacePDC(status: PDCStatus): boolean {
  return status === PDCStatus.BOUNCED;
}

/**
 * Check if PDC can be withdrawn
 * Only RECEIVED or DUE status PDCs can be withdrawn
 */
export function canWithdrawPDC(status: PDCStatus): boolean {
  return status === PDCStatus.RECEIVED || status === PDCStatus.DUE;
}

/**
 * Check if PDC can be cancelled
 * Only RECEIVED status PDCs can be cancelled
 */
export function canCancelPDC(status: PDCStatus): boolean {
  return status === PDCStatus.RECEIVED;
}

/**
 * Check if PDC is in a final state (no more actions possible)
 */
export function isPDCFinalState(status: PDCStatus): boolean {
  return [
    PDCStatus.CLEARED,
    PDCStatus.CANCELLED,
    PDCStatus.REPLACED,
    PDCStatus.WITHDRAWN
  ].includes(status);
}

/**
 * Format currency as AED for PDCs
 */
export function formatPDCCurrency(amount: number): string {
  return new Intl.NumberFormat('en-AE', {
    style: 'currency',
    currency: 'AED',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2
  }).format(amount);
}

/**
 * Calculate days until cheque date
 * Returns negative if past, positive if future, 0 if today
 */
export function getDaysUntilChequeDate(chequeDate: string): number {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const cheque = new Date(chequeDate);
  cheque.setHours(0, 0, 0, 0);
  const diffTime = cheque.getTime() - today.getTime();
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
}

/**
 * Check if cheque date is within due window (7 days)
 */
export function isWithinDueWindow(chequeDate: string): boolean {
  const days = getDaysUntilChequeDate(chequeDate);
  return days >= 0 && days <= 7;
}

/**
 * Format masked bank account number
 */
export function formatMaskedAccountNumber(accountNumber: string): string {
  if (!accountNumber || accountNumber.length < 4) {
    return '**** **** **** ****';
  }
  const lastFour = accountNumber.slice(-4);
  return `**** **** **** ${lastFour}`;
}

// ============================================================================
// LABEL RECORDS FOR DISPLAY
// ============================================================================

/**
 * PDC status labels for display
 */
export const PDC_STATUS_LABELS: Record<PDCStatus, string> = {
  [PDCStatus.RECEIVED]: 'Received',
  [PDCStatus.DUE]: 'Due',
  [PDCStatus.DEPOSITED]: 'Deposited',
  [PDCStatus.CLEARED]: 'Cleared',
  [PDCStatus.BOUNCED]: 'Bounced',
  [PDCStatus.CANCELLED]: 'Cancelled',
  [PDCStatus.REPLACED]: 'Replaced',
  [PDCStatus.WITHDRAWN]: 'Withdrawn'
};

/**
 * New payment method labels for display
 */
export const NEW_PAYMENT_METHOD_LABELS: Record<NewPaymentMethod, string> = {
  [NewPaymentMethod.BANK_TRANSFER]: 'Bank Transfer',
  [NewPaymentMethod.CASH]: 'Cash',
  [NewPaymentMethod.NEW_CHEQUE]: 'New Cheque'
};
